maintenance_request_time_performance/models/maintenance.py

```diff
@@ -0,0 +1,65 @@
+from odoo import api, models, fields, _
+from odoo.exceptions import ValidationError
+
+
+class MaintenanceStage(models.Model):
+    _inherit = 'maintenance.stage'
+
+    in_progress = fields.Boolean(
+        'In Progress',
+        default=False)
+
+    @api.model
+    def update_default_stage_in_progress(self):
+        stage = self.env.ref('maintenance.stage_1')
+        if stage:
+            stage.write({'in_progress': True})
+
+    @api.constrains('in_progress', 'done')
+    def check_maintenance_stage(self):
+        if self.in_progress and self.done:
+            raise ValidationError(
+                _("It can´t be in progress and done at the same time"))
+
+
+class MaintenanceRequest(models.Model):
+    _inherit = 'maintenance.request'
+
+    start_task_datetime = fields.Datetime(
+        string='Start Date and time of maintenance task completion')
+    end_task_datetime = fields.Datetime(
+        string='End Date and time of maintenance task completion')
+    real_task_duration = fields.Float(
+        compute='_compute_real_task_duration',
+        string='Real Duration',
+        store=True)
+
+    @api.depends('start_task_datetime', 'end_task_datetime')
+    def _compute_real_task_duration(self):
+        for request in self:
+            if request.end_task_datetime and not request.start_task_datetime:
+                raise ValidationError(
+                    _("It can´t be End task completion"
+                      " without Start task completion"))
+            total_hours = 0
+            if request.start_task_datetime and request.end_task_datetime \
+                    and request.start_task_datetime < request.end_task_datetime:
+                diff = request.end_task_datetime - request.start_task_datetime
+                total_hours = diff.total_seconds() / 3600
+            request.real_task_duration = total_hours
+
+    @api.constrains('stage_id')
+    def check_maintenance_request_stage(self):
+        for request in self:
+            if request.stage_id:
+                if request.stage_id.done and not request.start_task_datetime:
+                    raise ValidationError(
+                        _("It can´t be done without Start task completion"))
+                # if stage_id changes to in_progress -> fill start_task_datetime
+                if request.stage_id.in_progress \
+                        and not request.start_task_datetime:
+                    request.start_task_datetime = fields.Datetime.now()
+                # if stage_id changes to done stage -> fill end_task_datetime
+                if request.stage_id.done and not request.end_task_datetime:
+                    request.end_task_datetime = fields.Datetime.now()
```

Comment:
Well, I finally agree with you, setting a field within an `@api.constrains` method is a bad procedure.

In other hand, I've tested start and end datetime fields and I found these errors:
- It's possible to change start datetime and set it after end datetime.
- It's possible to unset start datetime when end time is already filled.

So I make the following proposal:
- change `start_task_datetime` and `end_task_datetime` definition and make them with a compute method, `store=True` and `readonly=False`, so they can be computed but also editable.
- the compute method (e.g. `_compute_task_datetimes`) can be used, when `stage_id` is changed, to update that values when empty (this part can be recovered from `check_maintenance_request_stage` current method).
- define another method (e.g. `check_maintenance_request_task_datetimes`), when any of the task datetime fields is manually changed, to only cover the validation part of the `check_maintenance_request_stage` current method